AK/UFixedBigInt.h

```diff
@@ -17,825 +19,526 @@
     }
 
-    // FIXME: subc for sizeof(T) < sizeof(U)
-    template<Unsigned U>
-    requires(sizeof(T) >= sizeof(U)) constexpr R subc(U const& other, bool& carry) const
-    {
-        bool low_carry = (!m_low && carry) || (m_low - carry) < other;
-        bool high_carry = !m_high && low_carry;
-
-        T lower = m_low - other - carry;
-        T higher = m_high - low_carry;
-        carry = high_carry;
-
-        return { lower, higher };
-    }
-    constexpr R subc(R const& other, bool& carry) const
+    constexpr UnsignedStorageReadonlySpan span() const
     {
-        bool low_carry = (!m_low && carry) || (m_low - carry) < other.low();
-        bool high_carry = (!m_high && low_carry) || (m_high - low_carry) < other.high();
-
-        T lower = m_low - other.low() - carry;
-        T higher = m_high - other.high() - low_carry;
-        carry = high_carry;
-
-        return { lower, higher };
+        return { m_data.data(), static_size };
     }
 
-    constexpr R operator+(bool const& other) const
-    {
-        bool carry = false; // unused
-        return addc((u8)other, carry);
-    }
-    template<Unsigned U>
-    constexpr R operator+(U const& other) const
+    // Binary utils
+    constexpr size_t popcnt() const
     {
-        bool carry = false; // unused
-        return addc(other, carry);
+        size_t result = 0;
+        for (size_t i = 0; i < m_data.size(); ++i)
+            result += popcount(m_data[i]);
+        return result;
     }
 
-    constexpr R operator-(bool const& other) const
+    constexpr size_t ctz() const
     {
-        bool carry = false; // unused
-        return subc((u8)other, carry);
+        size_t result = 0;
+        for (size_t i = 0; i < m_data.size(); ++i) {
+            if (m_data[i]) {
+                result += count_trailing_zeroes(m_data[i]);
+                break;
+            } else {
+                result += 8 * sizeof(m_data[i]);
+            }
+        }
+        return result;
     }
 
-    template<Unsigned U>
-    constexpr R operator-(U const& other) const
+    constexpr size_t clz() const
     {
-        bool carry = false; // unused
-        return subc(other, carry);
+        size_t result = 0;
+        for (size_t i = m_data.size(); i--;) {
+            if (m_data[i]) {
+                result += count_leading_zeroes(m_data[i]);
+                break;
+            } else {
+                result += 8 * sizeof(m_data[i]);
+            }
+        }
+        return result + bit_size - 8 * sizeof(NativeWord) * static_size;
     }
 
-    template<Unsigned U>
-    constexpr R& operator+=(U const& other)
-    {
-        *this = *this + other;
-        return *this;
-    }
-    template<Unsigned U>
-    constexpr R& operator-=(U const& other)
+    // Comparisons
+    constexpr bool operator!() const
     {
-        *this = *this - other;
-        return *this;
+        bool result = true;
+        for (size_t i = 0; i < m_data.size(); ++i)
+            result &= !m_data[i];
```

Comment:
I can't say which one of these outputs is better, and performance will surely depend on the exact input (but I prefer my version because it doesn't have jumps) https://godbolt.org/z/rshrfnzG9

Reply:
hmm… definitely something that canʼt be tackled comprehensively here—as you say, the exact architecture implementations and their splitting points is a matter of future…! 🙇🏻‍♂️